### PR TITLE
Cache checksum computation at an collection-element level

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Speedometer traces showed some expenses here.  What can happen is you might have a `List of references` A that we compute a checksum against.  But then if that list changes, but contains *almost* the entire set of references, we still recompute for the ones that are unchanged.  This keeps around the checksums for the unchanged items.

Also, we switch to consistently using a single mechanism to determine the mvid for PERef and ANalyzerReference.  This avoids unnecessarily additional costs that exist to get teh full metadata info, that we don't need for checksum computation.

Build: https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build/results?buildId=10440954&view=results

### CODE REVIEW

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Review by dibarbet — APPROVED** [review]
I think it looks fine, but might want to run through speedometer to check

File: src/Workspaces/Core/Portable/Serialization/SerializerService_Reference.cs [hunk 1]
```diff
@@ -195,7 +195,16 @@ private static Checksum CreatePortableExecutableReferenceChecksum(PortableExecut
         using (var writer = new ObjectWriter(stream, leaveOpen: true))
         {
             WritePortableExecutableReferencePropertiesTo(reference, writer);
-            WriteMvidsTo(TryGetMetadata(reference), writer);
+            if (Path.IsPathRooted(reference.FilePath))
```
[the remark this answers is not in the document]
Comment (ToddGrun) [reply src/Workspaces/Core/Portable/Serialization/SerializerService_Reference.cs:198]: IsPathRooted is just a string operation, right?

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Serialization/SerializerService_Reference.cs:198]: correct.

[the remark this answers is not in the document]
Comment (ToddGrun) [reply src/Workspaces/Core/Portable/Serialization/SerializerService_Reference.cs:198]: Is this waiting on a speedometer run?

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Serialization/SerializerService_Reference.cs:198]: yes.

**Pull request merged**